services/drupal/drupal_server.py

```diff
@@ -0,0 +1,105 @@
+# -*- coding: utf-8 -*-
+"""A Drupal CMS server based on Python's HTTPServer."""
+from __future__ import unicode_literals
+
+import os
+
+try:
+    from socketserver import ThreadingMixIn  # Python 3 exclusive
+    from http.server import HTTPServer, SimpleHTTPRequestHandler
+    from urllib.parse import unquote, urlparse
+except Exception:
+    from SocketServer import ThreadingMixIn  # Python 2 exclusive
+    from BaseHTTPServer import HTTPServer
+    from SimpleHTTPServer import SimpleHTTPRequestHandler
+    from urllib import unquote
+    from urlparse import urlparse
+
+
+WEB_PORT = 80
+WWW_FOLDER_NAME = "html"
+WEB_ALERT_TYPE_NAME = "drupal_rce"
+DEFAULT_SERVER_VERSION = "Apache 2"
+ALERTS = [WEB_ALERT_TYPE_NAME]
+
+
+class ThreadingHTTPServer(ThreadingMixIn, HTTPServer):
+    """Extend both classes to have threading capabilities."""
+
+    pass
+
+
+class HoneyHTTPRequestHandler(SimpleHTTPRequestHandler, object):
+    """Filter requests to catch Drupalgeddon 2 exploit attempts."""
+
+    def version_string(self):
+        """Return the web server name that we run on."""
+        return DEFAULT_SERVER_VERSION
+
+    def verify(self, query):
+        """Filter HTTP request to make sure it's not an exploit attempt."""
+        self.logger.debug("Query: " + query)
+        if query and query.find("&") != -1:
+            query_components = {}
+            for param in query.split("&"):
+                if param.find("=") == -1:
+                    continue
+                else:
+                    key, value = param.split("=")
+                    query_components[key] = value
+
+            for component in query_components:
+                if component.find("[#") != -1 and len(component) > 1:
+                    self.alert(event_name=WEB_ALERT_TYPE_NAME,
+                               request=query,
+                               orig_ip=self.client_address[0],
+                               orig_port=self.client_address[1])
+                    break
+
+    def do_GET(self):
+        """Handle an HTTP GET request."""
+        query = unquote(urlparse(self.path).query)
+        self.verify(query)
+        super(HoneyHTTPRequestHandler, self).do_GET()
+
+    def do_POST(self):
+        """Handle an HTTP POST request."""
+        content_length = int(self.headers['Content-Length'])
+        post_data = unquote(self.rfile.read(content_length).decode())
+        self.verify(post_data)
+        super(HoneyHTTPRequestHandler, self).do_GET()
+
+    def log_request(self, code="-", size="-"):
+        """Log an incoming request."""
+        self.logger.debug("debug: {request}, code: {code}, size: {size}".format(request=self.requestline,
```

Comment:
check SimpleHTTPServer, I think it's better to use standard log format like W3 so people can easily parse

Reply:
what about this one?